### PR TITLE
Refactor window to class: Finances

### LINKED ISSUE #13773: Refactor window to class: Finances
See #13737 for more details.

### PR DESCRIPTION
**Pull request by Iikorni**
This is my first contribution, so I'm sorry if it's a bit all over the place. I tried to model this refactor after `Cheats.cpp`, so some of the layout might be a bit off.

Closes #13773.

### CODE REVIEW

File: src/openrct2-ui/windows/Finances.cpp [hunk 1]
```diff
@@ -83,55 +81,63 @@ enum
...
-    MakeWidget        ({  0,  43}, {530, 266}, WindowWidgetType::Resize,  WindowColour::Secondary),
-    TAB_WIDGETS,
+    MAIN_FINANCES_WIDGETS(STR_FINANCIAL_SUMMARY, RSW_OTHER_TABS, RSH_SUMMARY, WW_OTHER_TABS, WH_SUMMARY),
     MakeWidget        ({130,  50}, {391, 211}, WindowWidgetType::Scroll,  WindowColour::Secondary, SCROLL_HORIZONTAL              ),
     MakeSpinnerWidgets({ 64, 279}, { 97,  14}, WindowWidgetType::Spinner, WindowColour::Secondary, STR_FINANCES_SUMMARY_LOAN_VALUE), // NB: 3 widgets.
     WIDGETS_END,
 };
 
-static rct_widget _windowFinancesCashWidgets[] =
+static rct_widget window_finances_cash_widgets[] =
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:120]: stick to the old naming it was the most correct following the coding style. This comment goes for all of the static local globals.

File: src/openrct2-ui/windows/Finances.cpp [hunk 2]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
-#pragma endregion
+rct_window* WindowFinancesResearchOpen()
+{
+    auto* window = FocusOrCreateFinancesWindow();
+
+    if (window != nullptr)
+        window->SetPage(WINDOW_FINANCES_PAGE_RESEARCH);
+
+    return window;
+}
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:1079]: include a new line.

File: src/openrct2-ui/windows/Finances.cpp [hunk 3]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
+        DrawTabImage(dpi, WINDOW_FINANCES_PAGE_SUMMARY, SPR_TAB_FINANCES_SUMMARY_0);
+        DrawTabImage(dpi, WINDOW_FINANCES_PAGE_FINANCIAL_GRAPH, SPR_TAB_FINANCES_FINANCIAL_GRAPH_0);
+        DrawTabImage(dpi, WINDOW_FINANCES_PAGE_VALUE_GRAPH, SPR_TAB_FINANCES_VALUE_GRAPH_0);
+        DrawTabImage(dpi, WINDOW_FINANCES_PAGE_PROFIT_GRAPH, SPR_TAB_FINANCES_PROFIT_GRAPH_0);
+        DrawTabImage(dpi, WINDOW_FINANCES_PAGE_MARKETING, SPR_TAB_FINANCES_MARKETING_0);
+        DrawTabImage(dpi, WINDOW_FINANCES_PAGE_RESEARCH, SPR_TAB_FINANCES_RESEARCH_0);
     }
+};
+
+static FinancesWindow* FocusOrCreateFinancesWindow()
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:1061]: I don't really see the point of this function. Is this the style from another window?

File: src/openrct2-ui/windows/Finances.cpp [hunk 4]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
-    w = window_bring_to_front_by_class(WC_FINANCES);
-    if (w == nullptr)
+public:
+    void OnOpen() override
     {
-        w = WindowCreateAutoPos(WW_OTHER_TABS, WH_SUMMARY, _windowFinancesPageEvents[0], WC_FINANCES, WF_10);
-        w->number = 0;
-        w->frame_no = 0;
-
+        SetPage(WINDOW_FINANCES_PAGE_SUMMARY);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:273]: _lastPaintedMonth should be set to `std::numeric_limits<int32_t>::max()` to force it to be recalculated on first draw.

**Comment by duncanspumpkin** [comment]
In future try avoid making modifications to the logic whilst you are doing these refactors to class as it makes reviewing much harder. Just do a follow on pr.

File: src/openrct2-ui/windows/Finances.cpp [hunk 5]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
     {
-        // Darken every even row
-        if (i % 2 == 0)
-            gfx_fill_rect(
-                dpi,
-                { screenCoords - ScreenCoordsXY{ 0, 1 }, screenCoords + ScreenCoordsXY{ row_width, (TABLE_CELL_HEIGHT - 2) } },
-                ColourMapA[w->colours[1]].lighter | 0x1000000);
-
-        screenCoords.y += TABLE_CELL_HEIGHT;
+        auto ft = Formatter::Common();
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:547]: Can you add a comment that this is setting loan widgets format arguments.

As this is done via a global it is very important that nothing else uses the global format arguments until the drawing is complete.

File: src/openrct2-ui/windows/Finances.cpp [hunk 6]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
         }
-        case WIDX_LOAN_DECREASE:
-            if (gBankLoan > 0)
+
+        return {};
+    }
+
+    void OnScrollDraw(int32_t scrollIndex, rct_drawpixelinfo& dpi) override
+    {
+        if (page == WINDOW_FINANCES_PAGE_SUMMARY)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:400]: return early so this isn't all indented.

File: src/openrct2-ui/windows/Finances.cpp [hunk 7]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
+#pragma endregion
 
-static void WindowFinancesSetPressedTab(rct_window* w)
-{
-    int32_t i;
-    for (i = 0; i < WINDOW_FINANCES_PAGE_COUNT; i++)
-        w->pressed_widgets &= ~(1ULL << (WIDX_TAB_1 + i));
-    w->pressed_widgets |= 1LL << (WIDX_TAB_1 + w->page);
-}
+    void InitializeScrollPosition(rct_widgetindex widgetIndex, int32_t scrollId)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:1025]: spelling mistake `InitialiseScrollPosition`

File: src/openrct2-ui/windows/Finances.cpp [hunk 8]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
+    void OnMouseDownResearch(rct_widgetindex widgetIndex)
     {
-        gDropdownItemsFormat[i] = STR_DROPDOWN_MENU_LABEL;
-        gDropdownItemsArgs[i] = ResearchFundingLevelNames[i];
-    }
-    WindowDropdownShowTextCustomWidth(
-        { w->windowPos.x + dropdownWidget->left, w->windowPos.y + dropdownWidget->top }, dropdownWidget->height() + 1,
-        w->colours[1], 0, Dropdown::Flag::StayOpen, 4, dropdownWidget->width() - 3);
+        rct_widget* dropdownWidget;
+        int32_t i;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:944]: Move this to the for loop as its only used there.

File: src/openrct2-ui/windows/Finances.cpp [hunk 9]
```diff
@@ -341,1048 +260,820 @@ static constexpr const uint32_t WindowFinancesPageHoldDownWidgets[] =
...
-    for (i = 0; i < 4; i++)
+    void OnMouseDownResearch(rct_widgetindex widgetIndex)
     {
-        gDropdownItemsFormat[i] = STR_DROPDOWN_MENU_LABEL;
-        gDropdownItemsArgs[i] = ResearchFundingLevelNames[i];
-    }
-    WindowDropdownShowTextCustomWidth(
-        { w->windowPos.x + dropdownWidget->left, w->windowPos.y + dropdownWidget->top }, dropdownWidget->height() + 1,
-        w->colours[1], 0, Dropdown::Flag::StayOpen, 4, dropdownWidget->width() - 3);
+        rct_widget* dropdownWidget;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Finances.cpp:943]: no need to declare this early.

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
Some minor comments but all looks about there.

**Comment by Broxzier** [comment]
@Iikorni Are you planning to continue working on this, and would you like some help?

**Comment by Iikorni** [comment]
@Broxzier My apologies for the late reply - I am still working on this, I just have been far too busy at work to devote the time necessary to fix this up. It will be done soon, hopefully.

**Comment by Broxzier** [comment]
No worries! If you are getting stuck or need some help, just ask either here or on Discord.

**Comment by Iikorni** [comment]
I am indeed working on finishing this up now, and it should be done within the next couple of days. Sorry that it's taken so long!

**Comment by tupaschoal** [comment]
@Iikorni sorry it took us a while, can you go over the line notes that Duncan made and mark them as resolved if you followed exactly the suggestion laid out there?

That makes it easier for us to know where to follow up from.
Thanks!

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**